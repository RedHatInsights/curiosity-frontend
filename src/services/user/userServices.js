import Cookies from 'js-cookie';
import LocaleCode from 'iso-639-1';
import { serviceCall } from '../config';
import { helpers } from '../../common';

/**
 * User, and RHSM, calls for locale and Opt-In.
 *
 * @memberof User
 * @module UserServices
 */

/**
 * ToDo: Review moving the getLocale function under platformServices.
 * Also review using window.navigator.language as the primary pull for language.
 */
/**
 * Return a browser locale, or fallback towards the platform locale cookie
 *
 * @returns {Promise<*>}
 */
const getLocale = () => {
  const defaultLang = {
    value: helpers.UI_LOCALE_DEFAULT,
    key: helpers.UI_LOCALE_DEFAULT_DESC
  };
  const parseLang = value => {
    const key = (value && LocaleCode.getName(value.split('-')[0])) || null;
    return (key && { value, key }) || undefined;
  };

  return serviceCall({
    url: async () => {
      const cookieLang = await (Cookies.get(process.env.REACT_APP_CONFIG_SERVICE_LOCALES_COOKIE) || '').replace(
        '_',
        '-'
      );

      return parseLang(cookieLang) || defaultLang;
    }
  });
};

/**
 * @apiMock {DelayResponse} 2000
 * @api {delete} /api/rhsm-subscriptions/v1/opt-in
 * @apiDescription Delete a RHSM account opt-in config
 *
 * Reference [RHSM API](https://github.com/RedHatInsights/rhsm-subscriptions/blob/main/api/rhsm-subscriptions-api-v1-spec.yaml)
 *
 * @apiSuccessExample {text} Success-Response:
 *     HTTP/1.1 204 OK
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 400 Bad Request
 *     {
 *        "errors": [
 *          {
 *            "status": "string",
 *            "code": "string",
 *            "title": "string",
 *            "detail": "string"
 *          }
 *        ]
 *     }
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 403 Forbidden
 *     {
 *        "errors": [
 *          {
 *            "status": "string",
 *            "code": "string",
 *            "title": "string",
 *            "detail": "string"
 *          }
 *        ]
 *     }
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 404 Not Found
 *     {
 *        "errors": [
 *          {
 *            "status": "string",
 *            "code": "string",
 *            "title": "string",
 *            "detail": "string"
 *          }
 *        ]
 *     }
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 500 Internal Server Error
 *     {
 *        "errors": [
 *          {
 *            "status": "string",
 *            "code": "string",
 *            "title": "string",
 *            "detail": "string"
 *          }
 *        ]
 *     }
 */
/**
 * Delete a RHSM account opt-in config.
 *
 * @returns {Promise<*>}
 */
const deleteAccountOptIn = () =>
  serviceCall({
    method: 'delete',
    url: process.env.REACT_APP_SERVICES_RHSM_OPTIN
  });

/**
 * @api {get} /api/rhsm-subscriptions/v1/opt-in
 * @apiDescription Get a RHSM account opt-in config
 *
 * Reference [RHSM API](https://github.com/RedHatInsights/rhsm-subscriptions/blob/main/api/rhsm-subscriptions-api-v1-spec.yaml)
 *
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *     {
 *        "data": {
 *          "opt_in_complete": true,
 *          "account": {
 *            "account_number": 12345,
 *            "tally_sync_enabled": true,
 *            "tally_reporting_enabled": true,
 *            "opt_in_type": "API",
 *            "created": "2017-08-04T17:32:05Z",
 *            "last_updated": "2017-08-04T17:32:05Z"
 *          },
 *          "org": {
 *            "org_id": 1111,
 *            "conduit_sync_enabled": true,
 *            "opt_in_type": "API",
 *            "created": "2017-08-04T17:32:05Z",
 *            "last_updated": "2017-08-04T17:32:05Z"
 *          }
 *        },
 *        "meta": {
 *          "account_number": 12345,
 *          "org_id": 1111
 *        }
 *     }
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 500 Internal Server Error
 *     {
 *        "errors": [
 *          {
 *            "status": "string",
 *            "code": "string",
 *            "title": "string",
 *            "detail": "string"
 *          }
 *        ]
 *     }
 */
/**
 * Get a RHSM account opt-in config.
 *
 * @param {object} options
 * @param {boolean} options.cancel
 * @param {string} options.cancelId
 * @returns {Promise<*>}
 */
const getAccountOptIn = (options = {}) => {
  const { cancel = true, cancelId } = options;
  return serviceCall({
    url: process.env.REACT_APP_SERVICES_RHSM_OPTIN,
    cancel,
    cancelId
  });
};

/**
 * @apiMock {DelayResponse} 1000
 * @apiMock {ForceStatus} 200
 * @api {put} /api/rhsm-subscriptions/v1/opt-in
 * @apiDescription Create/Update an account's opt-in configuration. Account and Org ID are defined by
 * the identity header. If no parameters are specified, everything will be enabled.
 *
 * Reference [RHSM API](https://github.com/RedHatInsights/rhsm-subscriptions/blob/main/api/rhsm-subscriptions-api-v1-spec.yaml)
 *
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *     {
 *        "data": {
 *          "opt_in_complete": true,
 *          "account": {
 *            "account_number": 12345,
 *            "tally_sync_enabled": true,
 *            "tally_reporting_enabled": true,
 *            "opt_in_type": "API",
 *            "created": "2017-08-04T17:32:05Z",
 *            "last_updated": "2017-08-04T17:32:05Z"
 *          },
 *          "org": {
 *            "org_id": 1111,
 *            "conduit_sync_enabled": true,
 *            "opt_in_type": "API",
 *            "created": "2017-08-04T17:32:05Z",
 *            "last_updated": "2017-08-04T17:32:05Z"
 *          }
 *        },
 *        "meta": {
 *          "account_number": 12345,
 *          "org_id": 1111
 *        }
 *     }
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 500 Internal Server Error
 *     {
 *        "errors": [
 *          {
 *            "status": "string",
 *            "code": "string",
 *            "title": "string",
 *            "detail": "string"
 *          }
 *        ]
 *     }
 */
/**
 * Update a RHSM account opt-in config.
 *
 * @param {object} params Query/search params
 * @returns {Promise<*>}
 */
const updateAccountOptIn = (params = {}) =>
  serviceCall({
    method: 'put',
    url: process.env.REACT_APP_SERVICES_RHSM_OPTIN,
    params
  });

const userServices = { getLocale, deleteAccountOptIn, getAccountOptIn, updateAccountOptIn };

/**
 * Expose services to the browser's developer console.
 */
helpers.browserExpose({ userServices });

export { userServices as default, userServices, getLocale, deleteAccountOptIn, getAccountOptIn, updateAccountOptIn };
