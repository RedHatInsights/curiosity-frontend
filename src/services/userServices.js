import Cookies from 'js-cookie';
import LocaleCode from 'locale-code';
import _isPlainObject from 'lodash/isPlainObject';
import { getUser, getUserPermissions } from './platformServices';
import { serviceCall } from './config';
import { helpers } from '../common';

/**
 * Apply an emulated API response to the platforms getUser method.
 *
 * @returns {Promise<{data: {permissions: (void|*[]), user: void}, message: string, status: number}>}
 */
const authorizeUser = async () => {
  let message = '{ auth.getUser, getUserPermissions } = insights.chrome';
  let userData;
  let userPermissions;

  try {
    userData = await getUser();
    userPermissions = await getUserPermissions(helpers.UI_NAME);
  } catch (e) {
    message = e.message;
  }

  if (_isPlainObject(userData) && Object.keys(userData).length) {
    return Promise.resolve({ data: { user: userData, permissions: userPermissions || [] }, message, status: 200 });
  }

  const emulatedErrorResponse = {
    ...new Error(message),
    message,
    status: 418
  };

  return Promise.reject(emulatedErrorResponse);
};

/**
 * Return a platform locale value from a cookie.
 *
 * @private
 * @returns {{value: string, key: string | null}|null}
 */
const getLocaleFromCookie = () => {
  const value = (Cookies.get(process.env.REACT_APP_CONFIG_SERVICE_LOCALES_COOKIE) || '').replace('_', '-');
  const key = (value && LocaleCode.getLanguageName(value)) || null;

  return (key && { value, key }) || null;
};

/**
 * Return platform locale.
 *
 * @returns {Promise<{data: void}>}
 */
const getLocale = () => {
  const defaultLocale = {
    value: helpers.UI_LOCALE_DEFAULT,
    key: helpers.UI_LOCALE_DEFAULT_DESC
  };

  return new Promise(resolve =>
    resolve({
      data: getLocaleFromCookie() || defaultLocale
    })
  );
};

const logoutUser = () =>
  new Promise(resolve => {
    resolve({});
  });

/**
 * @apiMock {DelayResponse} 2000
 * @api {delete} /api/rhsm-subscriptions/v1/opt-in
 * @apiDescription Delete a RHSM account opt-in config
 *
 * Reference [RHSM API](https://github.com/RedHatInsights/rhsm-subscriptions/blob/master/api/rhsm-subscriptions-api-spec.yaml)
 *
 * @apiSuccessExample {text} Success-Response:
 *     HTTP/1.1 204 OK
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 400 Bad Request
 *     {
 *        "errors": [
 *          {
 *            "status": "string",
 *            "code": "string",
 *            "title": "string",
 *            "detail": "string"
 *          }
 *        ]
 *     }
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 403 Forbidden
 *     {
 *        "errors": [
 *          {
 *            "status": "string",
 *            "code": "string",
 *            "title": "string",
 *            "detail": "string"
 *          }
 *        ]
 *     }
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 404 Not Found
 *     {
 *        "errors": [
 *          {
 *            "status": "string",
 *            "code": "string",
 *            "title": "string",
 *            "detail": "string"
 *          }
 *        ]
 *     }
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 500 Internal Server Error
 *     {
 *        "errors": [
 *          {
 *            "status": "string",
 *            "code": "string",
 *            "title": "string",
 *            "detail": "string"
 *          }
 *        ]
 *     }
 */
/**
 * Delete a RHSM account opt-in config.
 *
 * @returns {Promise<*>}
 */
const deleteAccountOptIn = () =>
  serviceCall({
    method: 'delete',
    url: process.env.REACT_APP_SERVICES_RHSM_OPTIN
  });

/**
 * @api {get} /api/rhsm-subscriptions/v1/opt-in
 * @apiDescription Get a RHSM account opt-in config
 *
 * Reference [RHSM API](https://github.com/RedHatInsights/rhsm-subscriptions/blob/master/api/rhsm-subscriptions-api-spec.yaml)
 *
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *     {
 *        "data": {
 *          "opt_in_complete": true,
 *          "account": {
 *            "account_numner": 12345,
 *            "tally_sync_enabled": true,
 *            "tally_reporting_enabled": true,
 *            "opt_in_type": "API",
 *            "created": "2017-08-04T17:32:05Z",
 *            "last_updated": "2017-08-04T17:32:05Z"
 *          },
 *          "org": {
 *            "org_id": 1111,
 *            "conduit_sync_enabled": true,
 *            "opt_in_type": "API",
 *            "created": "2017-08-04T17:32:05Z",
 *            "last_updated": "2017-08-04T17:32:05Z"
 *          }
 *        },
 *        "meta": {
 *          "account_number": 12345,
 *          "org_id": 1111
 *        }
 *     }
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 500 Internal Server Error
 *     {
 *        "errors": [
 *          {
 *            "status": "string",
 *            "code": "string",
 *            "title": "string",
 *            "detail": "string"
 *          }
 *        ]
 *     }
 */
/**
 * Get a RHSM account opt-in config.
 *
 * @returns {Promise<*>}
 */
const getAccountOptIn = () =>
  serviceCall({
    url: process.env.REACT_APP_SERVICES_RHSM_OPTIN,
    cancel: true
  });

/**
 * @apiMock {DelayResponse} 1000
 * @apiMock {ForceStatus} 200
 * @api {put} /api/rhsm-subscriptions/v1/opt-in
 * @apiDescription Create/Update an account's opt-in configuration. Account and Org ID are defined by
 * the identity header. If no parameters are specified, everything will be enabled.
 *
 * Reference [RHSM API](https://github.com/RedHatInsights/rhsm-subscriptions/blob/master/api/rhsm-subscriptions-api-spec.yaml)
 *
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *     {
 *        "data": {
 *          "opt_in_complete": true,
 *          "account": {
 *            "account_numner": 12345,
 *            "tally_sync_enabled": true,
 *            "tally_reporting_enabled": true,
 *            "opt_in_type": "API",
 *            "created": "2017-08-04T17:32:05Z",
 *            "last_updated": "2017-08-04T17:32:05Z"
 *          },
 *          "org": {
 *            "org_id": 1111,
 *            "conduit_sync_enabled": true,
 *            "opt_in_type": "API",
 *            "created": "2017-08-04T17:32:05Z",
 *            "last_updated": "2017-08-04T17:32:05Z"
 *          }
 *        },
 *        "meta": {
 *          "account_number": 12345,
 *          "org_id": 1111
 *        }
 *     }
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 500 Internal Server Error
 *     {
 *        "errors": [
 *          {
 *            "status": "string",
 *            "code": "string",
 *            "title": "string",
 *            "detail": "string"
 *          }
 *        ]
 *     }
 */
/**
 * Update a RHSM account opt-in config.
 *
 * @param {object} params Query/search params
 * @returns {Promise<*>}
 */
const updateAccountOptIn = (params = {}) =>
  serviceCall({
    method: 'put',
    url: process.env.REACT_APP_SERVICES_RHSM_OPTIN,
    params
  });

const userServices = { authorizeUser, getLocale, logoutUser, deleteAccountOptIn, getAccountOptIn, updateAccountOptIn };

/**
 * Expose services to the browser's developer console.
 */
helpers.browserExpose({ userServices });

export {
  userServices as default,
  userServices,
  authorizeUser,
  getLocale,
  logoutUser,
  deleteAccountOptIn,
  getAccountOptIn,
  updateAccountOptIn
};
