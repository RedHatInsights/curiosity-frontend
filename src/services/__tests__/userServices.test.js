import userServices from '../userServices';

describe('UserServices', () => {
  it('should export a specific number of methods and classes', () => {
    expect(Object.keys(userServices)).toHaveLength(3);
  });

  it('should have specific methods', () => {
    expect(userServices.authorizeUser).toBeDefined();
    expect(userServices.getLocale).toBeDefined();
    expect(userServices.logoutUser).toBeDefined();
  });

  it('should return promises for every method', done => {
    const promises = Object.keys(userServices).map(value => userServices[value]());

    Promise.all(promises).then(success => {
      expect(success.length).toEqual(Object.keys(userServices).length);
      done();
    });
  });
});
