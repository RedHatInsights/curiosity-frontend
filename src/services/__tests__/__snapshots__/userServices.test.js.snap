// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserServices should return a failed authorized user on an unexpected response: failed authorized user: unexpected 1`] = `
Object {
  "message": "{ auth.getUser, getUserPermissions } = insights.chrome",
  "status": 418,
}
`;

exports[`UserServices should return a failed authorized user on empty response: failed authorized user: empty object 1`] = `
Object {
  "message": "{ auth.getUser, getUserPermissions } = insights.chrome",
  "status": 418,
}
`;

exports[`UserServices should return a failed authorized user: failed authorized user 1`] = `
Object {
  "message": "{ getUser } = insights.chrome.auth, insights.chrome.auth.getUser is not a function",
  "status": 418,
}
`;

exports[`UserServices should return a specific locale cookie value 1`] = `
Object {
  "data": Object {
    "key": "English",
    "value": "en-US",
  },
}
`;

exports[`UserServices should return a successful authorized user: success authorized user 1`] = `
Object {
  "data": Object {
    "permissions": Array [],
    "user": Object {
      "identity": Object {
        "account_number": "0",
        "type": "User",
      },
    },
  },
  "message": "{ auth.getUser, getUserPermissions } = insights.chrome",
  "status": 200,
}
`;

exports[`UserServices should return default locale if no locale cookie is present 1`] = `
Object {
  "data": Object {
    "key": "English",
    "value": "en-US",
  },
}
`;

exports[`UserServices should return the default locale with an invalid ISO_639 code 1`] = `
Object {
  "data": Object {
    "key": "English",
    "value": "en-US",
  },
}
`;
